sklearn/utils/tests/test_validation.py

```diff
@@ -728,8 +742,25 @@ def test_check_dataframe_warns_on_dtype():
     df_mixed_numeric = pd.DataFrame([[1., 2, 3], [4., 5, 6]])
     assert_warns(DataConversionWarning, check_array, df_mixed_numeric,
                  dtype='numeric', warn_on_dtype=True)
-    assert_no_warnings(check_array, df_mixed_numeric.astype(int),
-                       dtype='numeric', warn_on_dtype=True)
+
+
+def test_check_dataframe_no_warning_on_dtype():
+    pd = importorskip("pandas")
+    df = pd.DataFrame([[1, 2, 3], [4, 5, 6]], dtype=object)
+
+    with pytest.warns(None) as record:
+        warnings.simplefilter("ignore", DeprecationWarning)  # 0.23
```

Comment:
please don't use `warnings.simplefilter` and use instead

`@pytest.mark.filterwarnings('ignore: ...')`

Reply:
I cannot seem to get it to work using:

   ```
 with pytest.warns(None) as record:
        pytest.mark.filterwarnings("ignore: 'warn_on_dtype' is deprecated")  # 0.23
        check_array(df, dtype='object', warn_on_dtype=True)
    assert len(record) == 0
```

I am afraid `pytest.mark.filterwarnings` can only work as a decorator using a `@` at the definition of the method. And cannot be used as above.

Do you mean removing pytest.warns(None) from here completely and then using a decorator.

Also checking for something similar I found this:
https://github.com/pytest-dev/pytest/issues/4011

Reply:
I see. Like OP in your link (thanks for the ref!) I thought that using the decorator would take precedence over the context manager. In this case I think that your solution is correct. And you should be able to put these two checks back where they were before (i.e. in `test_check_dataframe_warns_on_dtype`).

Apologies for the extra work

Reply:
Done. No problem, thank you for reviewing it.